dhcpv4/nclient4/client.go

```diff
@@ -162,6 +162,18 @@ type Client struct {
 	// TransactionID. receiveLoop uses this map to determine which channel
 	// to send a new DHCP message to.
 	pending map[dhcpv4.TransactionID]*pendingCh
+
+	//clientIdOptions is a list of DHCPv4 option code that DHCP server used to
+	//identify client other than the HWAddress,
+	//like client-id, option82/remote-id..etc
+	clientIDOptions dhcpv4.OptionCodeList
+
+	//lease info after DORA, nil before DORA
+	lease *DHCPv4ClientLease
```

Comment:
I don't think the current lease should be stored in this client, and I don't think this client should be concerned with applying the lease information to the host system.

I'd rather a user be able to make the choice of what to store and where for themselves, and that a user be able to make the choice of what configuration to apply to the host system and how, because I have a feeling it's going to be different for some.

I think FB tends to use the netboot package in this library, and we tend to use https://github.com/u-root/u-root/tree/master/pkg/dhclient, but either way, I don't think it belongs in nclient4.

reasoning being I'd rather not support this for all users if we cannot get every edge case correct, which neither netboot nor u-root's pkg/dhclient nor this currently do. and if we do, then it'd definitely be in a separate package from nclient4, just to decouple it -- i don't want u-root, for example, to be saddled with this dependency when it already has its own implementation of this.

My overall recommendation would be to keep the RenewalTime and RebindingTime -- that's really useful -- and the Release function. But let users pass in the lease they want released, rather than using one stored in this client:

```go
func (c *Client) Release(ack *DHCPv4) error {

}
```

That `ack` should have all the information needed to assemble a `Release`. This would be similar to the way that `Advertise` and `Request` interact in this client already.

Reply:
ack does not necessarily have all the information needed to build release, as my comments above, in case DHCP sever uses DHCP option to identify a DHCP client, for example, client-id option, ack will not have the client-id option;
I agree that this library shouldn't force how user apply the lease, that's why there is WithApplyLeaseHandler(h func(DHCPv4ClientLease, bool) error) for user to provide their own handler to however applying lease as they like;  the default handler (defaultLeaseApplyHandler)  only add assigned address to the interface, so that release message could use it as the source address, and if user doesn't want that behavoir, it could simply add and empty handler via  WithApplyLeaseHandler

Reply:
I think part of not forcing it, though, means not forcing other users of the library to take on the netlink dependency. E.g. the two different users of this library use different netlink libraries to achieve their goals, and both users care a lot about binary size requirements. We don't want both libraries in our dependencies, we want to choose. Unfortunately I don't trust Go's dead code elimination enough to be sure it'll be eliminated when it's not used.

Reply:
I'm open to committing something like u-root's pkg/dhclient in this library as well, for convenience, if it's maintainable. Some package that's higher-level than nclient4, but not in nclient4 imho.

Reply:
for netlink library, I could switch to  jsimonetti/rtnetlink, it shouldn't be difficult if that's your only concern;
however if you think nclient4 should only be low-level (e.g. remain stateless, no need for lease/release), then maybe then I just have to create another client ...

Reply:
I don't have a preference for one or the other, but we do need to satisfy requirements for *both* users of this library, which includes not having unwanted dependency. vishv/netlink is unwanted for those that use rtnetlink, and rtnetlink is unwanted for those who still use vishv/netlink. A wrapper around this client for your use case would be preferred IMHO. @pmazzini, wdyt?

Reply:
> A wrapper around this client for your use case would be preferred IMHO.

Yes, I agree. Maybe even a separate package inside this repo.